src/internal/log/http.go

```diff
@@ -44,7 +46,43 @@ func AddLoggerToHTTPServer(rctx context.Context, name string, s *http.Server) {
 			ctx = ChildLogger(ctx, "", WithFields(id))
 			ctx = metadata.NewOutgoingContext(ctx, metadata.MD{"x-request-id": requestID})
 			r = r.WithContext(ctx)
+
+			var respStatusCode int
+			var errbuf []byte
+			w = httpsnoop.Wrap(w, httpsnoop.Hooks{
+				WriteHeader: func(next httpsnoop.WriteHeaderFunc) httpsnoop.WriteHeaderFunc {
+					return func(code int) {
+						respStatusCode = code
+						next(code)
+					}
+				},
+				Write: func(next httpsnoop.WriteFunc) httpsnoop.WriteFunc {
+					return func(b []byte) (int, error) {
+						if respStatusCode == 0 {
+							respStatusCode = 200
+						}
+						if respStatusCode >= 400 {
+							if errbuf == nil {
+								// TODO would it be better to statically alloc this? (e.g.
+								// [1000]byte above?)
+								errbuf = make([]byte, 0, 1000)
+							}
+							if len(errbuf) < cap(errbuf) {
+								m := len(errbuf)
+								n := min(len(b), cap(errbuf)-m)
+								errbuf = errbuf[:m+n] // extend errbuf
+								copy(errbuf[m:], b[:n])
+							}
+						}
+						return next(b)
+					}
+				},
+			})
+
 			orig.ServeHTTP(w, r)
+			if ua := r.Header.Get("user-agent"); ua != "Envoy/HC" {
+				Info(ctx, "http response", zap.String("method", r.Method), zap.String("host", r.Host), zap.String("path", url), zap.String("peer", r.RemoteAddr), id, zap.Int("status-code", respStatusCode), zap.ByteString("error-msg", errbuf))
```

Comment:
Up to you if you want method/host/path/peer here.  They are logged with the request, and with a logger that generates a unique ID for each request.  So you can find the ID of this message, look for the request with the same ID, and then get that information.

For `status-code` we use camelCase throughout, but I think we can just call it `code`.  For the error message, you aren't testing anywhere if the response type is `text/plain; charset=utf-8`, so it could be binary data, which would be ugly in the logs.  I think `zap.Binary` is safer; yielding base64 encoded binary.  The called code should `log.Info` the error message it wants to serve anyway, in case the request aborts before it ends up being written to the ResponseWriter.

Reply:
It would also be nice to have a total number of bytes sent here.  (After errbuf fills up, keep maintaining a counter and log it here.)